server/statistics/statistics.go

```diff
@@ -0,0 +1,138 @@
+package statistics
+
+import (
+	"sync"
+	"time"
+
+        "github.com/ubports/ubuntu-push/logger"
+
+)
+
+type Statistics struct {
+
+	logger logger.Logger
+	//Mutex to protect the statistics from concurrent updates
+        updating sync.Mutex
+
+        //registered devices per...
+        //5 mins
+        reg_devices_5min int32
+        //60 mins
+        reg_devices_60min int32
+        //1 day
+        reg_devices_1day int32
+        //7 days
+        reg_devices_7day int32
+
+        unicasts_5min uint32
+        unicasts_60min uint32
+        unicasts_1day uint32
+        unicasts_7day uint32
+
+        broadcasts_5min uint32
+        broadcasts_60min uint32
+        broadcasts_1day uint32
+        broadcasts_7day uint32
+
+}
+
+func NewStatistics(logger logger.Logger) *Statistics {
+
+	return &Statistics{
+		logger: logger,
+		updating: sync.Mutex{},
+	}
+
+}
+
+func (stats *Statistics) ResetAll() {
+
+	stats.Reset5min()
+	stats.Reset60min()
+	stats.Reset1day()
+	stats.Reset7day()
+}
+
+func (stats *Statistics) Reset5min() {
+
+	stats.reg_devices_5min = 0
+	stats.unicasts_5min = 0
+}
+
+func (stats *Statistics) Reset60min() {
+
+	stats.reg_devices_60min = 0
+	stats.unicasts_60min = 0
+}
+
+func (stats *Statistics) Reset1day() {
+
+	stats.reg_devices_1day = 0
+	stats.unicasts_1day = 0
+}
+
+func (stats *Statistics) Reset7day() {
+
+	stats.reg_devices_7day = 0
+	stats.unicasts_7day = 0
+}
+
+func (stats *Statistics) DecreaseDevices() {
+
+	stats.updating.Lock()
+	stats.reg_devices_5min--
+	stats.reg_devices_60min--
+	stats.reg_devices_1day--
+	stats.reg_devices_7day--
+	stats.updating.Unlock()
+}
+
+func (stats *Statistics) IncreaseDevices() {
+
+	stats.updating.Lock()
+	stats.reg_devices_5min++
+	stats.reg_devices_60min++
+	stats.reg_devices_1day++
+	stats.reg_devices_7day++
+	stats.updating.Unlock()
+}
+
+func (stats *Statistics) IncreaseUnicasts() {
+
+	stats.updating.Lock()
+	stats.unicasts_5min++
+	stats.unicasts_60min++
+	stats.unicasts_1day++
+	stats.unicasts_7day++
+	stats.updating.Unlock()
+}
+
+//Shall be called periodically (every 5 m,inutes) to output stats
+func (stats *Statistics) PrintStats() {
+
+	callcount := 0
+	t := time.NewTicker(time.Minute * 5)
+	for {
+		stats.updating.Lock()
+		stats.logger.Infof("")
+		stats.logger.Infof("        |  Devices   |  Unicasts  | Broadcasts |")
+		stats.logger.Infof("5 mins  | %10v | %10v | %10v |", stats.reg_devices_5min, stats.unicasts_5min, stats.broadcasts_5min)
+		stats.logger.Infof("60 mins | %10v | %10v | %10v |", stats.reg_devices_60min, stats.unicasts_60min, stats.broadcasts_60min)
+		stats.logger.Infof("1 day   | %10v | %10v | %10v |", stats.reg_devices_1day, stats.unicasts_1day, stats.broadcasts_1day)
+		stats.logger.Infof("7 days  | %10v | %10v | %10v |", stats.reg_devices_7day, stats.unicasts_7day, stats.broadcasts_7day)
+		callcount++
+		stats.Reset5min()
+		if callcount % 12 == 0 {
```

Comment:
This is wrong: you should not reset the timers, but rather purge the older devices (and for that, you need to keep track of the time when IncreaseDevices() was called. For example, suppose that a new device gets added every 6 minutes: the stats for "60 mins" should always report 10 devices (plus or minus one), while after one hour of activity your implementation would return 0.

Reply:
Its a different interpretation: This is not the total device count but how many devices were added or removed. Eg here:
![grafik](https://user-images.githubusercontent.com/16186470/34197766-2b9b4c38-e568-11e7-878d-d49296a7ac53.png)

as you see the value also can get negative. For me as a infrastructure guy its more important how much the server is being challenged with adding or removing devices, rather than the static number. But we can add this too

Reply:
Yes, but I'm afraid it doesn't change the fact that most of the data you print is wrong: the "5 mins" line is the only one correct.
Please follow my example: if one device is added every 6 minutes, and no one is removed, what should the  "60 mins" line print, after 65 minutes of server activity? 10 or 0?

Reply:
Ok now I got you ;) - well currently its not a rolling stats so it will get cleared at minute 60 and therefore will print 0 at minute 65. At minute 66, in your point of view, it should rather print 9 right? That would roll out the first device added at minute 0 lets say...

Reply:
Could you not use an expiring counter? something like this https://paste.ubuntu.com/26219933/ (note this is not tested) You could new up an statistics counter for each period you want to track and let the janitor expire stats as you go. THen you just need to increment each counter with an expiry time and call count() when you the running count. Just an idea...

Reply:
I think its much easier made with a ring typed element for all stats except the 5mins stats. Then, you just populate the ring with every tick, and sum up all elements. As soon as the ring gets full the last element will be evicted automatically.